src/theory/evaluator.cpp

```diff
@@ -140,30 +158,103 @@ EvalResult Evaluator::evalInternal(TNode n,
       continue;
     }
 
+    bool doProcess = true;
     bool doEval = true;
     for (const auto& currNodeChild : currNode)
     {
-      if (results.find(currNodeChild) == results.end())
+      itr = results.find(currNodeChild);
+      if (itr == results.end())
       {
         queue.emplace_back(currNodeChild);
+        doProcess = false;
+      }
+      else if (itr->second.d_tag == EvalResult::INVALID)
+      {
+        // we cannot evaluate since there was an invalid child
         doEval = false;
       }
     }
+    Trace("evaluator") << "Evaluator: visit " << currNode
+                       << ", process = " << doProcess
+                       << ", evaluate = " << doEval << std::endl;
 
-    if (doEval)
+    if (doProcess)
     {
       queue.pop_back();
 
       Node currNodeVal = currNode;
+
+      // The code below should either:
+      // (1) store a valid EvalResult into results[currNode], or
+      // (2) store an invalid EvalResult into results[currNode] and
+      // store the result of substitution + rewriting currNode { args -> vals }
+      // into evalAsNode[currNode].
+
+      // If we did not successfully evaluate all children
+      if (!doEval)
+      {
+        // Reconstruct the node with a combination of the children that
+        // successfully evaluated, and the children that did not.
+        Trace("evaluator") << "Evaluator: collect arguments" << std::endl;
+        std::vector<Node> echildren;
+        if (currNode.getMetaKind() == kind::metakind::PARAMETERIZED)
+        {
+          echildren.push_back(currNode.getOperator());
+        }
+        for (const auto& currNodeChild : currNode)
+        {
+          itr = results.find(currNodeChild);
+          if (itr->second.d_tag == EvalResult::INVALID)
+          {
+            // could not evaluate this child, look in the node cache
+            itn = evalAsNode.find(currNodeChild);
+            if (itn != evalAsNode.end())
+            {
+              // take cached value
+              echildren.push_back(itn->second);
+            }
+            else
+            {
+              // must take original
+              echildren.push_back(currNodeChild);
```

Comment:
Aren't we guaranteed that a child is either in `results` or `evalAsNode`?

Reply:
Yes, I can turn this into an Assert.